### PR TITLE
11 command parser

### PR DESCRIPTION
**Pull request by Bawfen**

### CODE REVIEW

**Comment by Bawfen** [comment]
Also resolves issue #28

File: src/main/java/seedu/duke/commands/CommandType.java [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+            throw new NullPointerException("Command is null");
+        }
+        for(CommandType commandType : CommandType.values()){
+            if(commandType.getCommand().equals(command)){
+                return commandType;
+            }
+        }
+        throw new IllegalArgumentException("No command found named "+command);
+    }
+}
```
Comment (woowenjun99) [inline src/main/java/seedu/duke/commands/CommandType.java:26]: If the commands are just capital letter and the dashes are replaced with underscore, do considering implementing it in such a manner:

```java
public enum CommandType {
    EXIT, SHOW_RATE;
}
```
## String to Enum
```java
String command = args[0].replaceAll("-", "_").toUpperCase();
CommandType type = CommandType.valueOf(command);
```
By default, the second line should already throw an error if no such enum value is found. Example can be found in the other file with the ShowRate class. You can create an utility class with generics for future enums.

## Enum To String
```java
String command = CommandType.EXIT.name().replace("_", "-").toLowerCase();
```
## Alternative
Alternatively, just use an interface.

File: src/main/java/seedu/duke/parser/Parser.java [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+import seedu.duke.commands.CommandType;
+import seedu.duke.commands.ExitCommand;
+import seedu.duke.commands.ShowRateCommand;
+public class Parser {
+    public static Command parseInput(String input) throws IndexOutOfBoundsException, IllegalArgumentException{
+        CommandType command;
+        String[] args = input.split(" ");
+
+        // trim whitespace and convert to uppercase so input matches the enum names
+        command = CommandType.get(args[0]);
```
Comment (woowenjun99) [inline src/main/java/seedu/duke/parser/Parser.java:13]: Consider declaring the variable when you need it as part of good coding standards.

File: src/main/java/seedu/duke/commands/ShowRateCommand.java [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+
+import seedu.duke.Currency;
+
+public class ShowRateCommand extends Command {
+    private Currency from;
+    private Currency to;
+    public ShowRateCommand(String from, String to) {
+        super(false);
+        this.from = Currency.valueOf(from);
+        this.to = Currency.valueOf(to);
```
Comment (woowenjun99) [inline src/main/java/seedu/duke/commands/ShowRateCommand.java:11]: These 2 lines throw an Exception when the currency does not exist. Do remember to catch the error prior to passing it in should the currency be invalid.

**Review by woowenjun99 — COMMENTED** [review]
Good job on the task. I have left some comments on the class, do have a look at it and let me know if further clarifications are required. On top of the this skeleton, please help to do the following:

- [ ] Write some JavaDocs
- [x] Carry out some unit testing. For the second comment regarding the `Currency.valueOf(from)`, if no testing is performed, the program might crash.

File: src/main/java/seedu/duke/commands/CommandType.java (hunk 1 above)
Comment (Bawfen) [reply src/main/java/seedu/duke/commands/CommandType.java:26]: What is the benefit to implementing it as generic commands which seem to be less readable as opposed to the current implementation?

File: src/main/java/seedu/duke/commands/ShowRateCommand.java (hunk 3 above)
Comment (Bawfen) [reply src/main/java/seedu/duke/commands/ShowRateCommand.java:11]: I will leave implementation of tests to the person who is fully implementing the command in #16

File: src/main/java/seedu/duke/parser/Parser.java (hunk 2 above)
Comment (Bawfen) [reply src/main/java/seedu/duke/parser/Parser.java]: In the latest revision, the CommandType.get is wrapped in a try-catch block so it is necessary to declare the variable first

File: src/main/java/seedu/duke/commands/ShowRateCommand.java (hunk 3 above)
Comment (woowenjun99) [reply src/main/java/seedu/duke/commands/ShowRateCommand.java]: Agreed

File: src/main/java/seedu/duke/commands/CommandType.java (hunk 1 above)
Comment (woowenjun99) [reply src/main/java/seedu/duke/commands/CommandType.java:26]: I see what you mean. Maybe a generic utils package is not necessary.

**Pull request merged**